torch/fx/passes/reinplace.py

```diff
@@ -0,0 +1,257 @@
+import torch
+import torch.fx as fx
+from torch.fx._compatibility import compatibility
+
+from torch.multiprocessing.reductions import StorageWeakRef
+
+from collections import defaultdict
+from typing import Iterable
+import itertools
+
+__all__ = ['reinplace']
+
+def concatMap(func, xs):
+    for x in xs:
+        for r in func(x):
+            yield r
+
+class GatherReinplaceDatastructures(torch.fx.Interpreter):
+    def __init__(self, gm):
+        super().__init__(gm)
+        self.node_counter = -1
+        # graph variable names of all program inputs
+        self.input_names = []
+        # Groups sets of the variable names of all aliased variables in the program
+        self.alias_map = defaultdict(set)
+        # For a given variable, tells you the names of all nodes in the program that use it.
+        # (including the index of that node, so we know where in the program it is)
+        self.names_to_used_nodes = defaultdict(list)
+        # When you call `x.copy_(y)`, you overwrite x's data.
+        # This stores the name of the node (if any) where a given variable in the program is overwritten.
+        # We need to track this information for program inputs
+        self.last_node_where_variable_is_overwritten = defaultdict(lambda: -1)
+
+    def run_node(self, node):
+        self.node_counter += 1
+        args, kwargs = self.fetch_args_kwargs_from_env(node)
+
+        # Update names_to_used_nodes map for this node
+        # copy_() doesn't read from its first argument; it writes to it, overwriting previous data.
+        # We don't want to treat it as "being used as an input".
+        node_args = node.args
+        if node.target is torch.ops.aten.copy_.default:
+            node_args = node_args[1:]
+
+        for arg in itertools.chain(node_args, node.kwargs.values()):
+            if not isinstance(arg, Iterable):
+                arg = [arg]
+            for a in arg:
+                if isinstance(a, fx.node.Node):
+                    # We map to the actual fx.Node object, because later we might need to
+                    # modify the arguments that this node is called with
+                    self.names_to_used_nodes[a.name].append([self.node_counter, node])
```

Comment:
It's probably safer to use tree_map for this; as written this code can only handle Tensor and list of Tensor, if a Tensor pops up anywhere else it will be missed.